modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Engine.scala

```diff
@@ -3,36 +3,374 @@
+      case EventSystem(se) => handleSystemEvent(se)
+    }) *> get
+  }
+
+  // Kudos to @tpolecat
+  /** Traverse a process with a stateful computation. */
+  private def mapEvalState[A, S, B](
+                                     fs: Process[Task, A], s0: S, f: A => StateT[Task, S, (B, Option[Process[Task, A]])]
+                                   ): Process[Task, B] = {
+    def go(fi: Process[Task, A], si: S): Process[Task, B] = {
+      Process.eval(fi.unconsOption).flatMap {
+        case None => Process.halt
+        case Some((h, t)) => Process.eval(f(h).run(si)).flatMap {
+          case (s, (b, p)) => Process.emit(b) ++ go(p.map(_ merge t).getOrElse(t), s)
+        }
+      }
+    }
+
+    go(fs, s0)
+  }
+
+  private def runE(ev: Event[D]): HandleP[(Event[D], Engine.State[D]), D] =
+    run(ev).map((ev, _))
+
+  def process(input: Process[Task, Event[D]])(qs: Engine.State[D]): Process[Task, (Event[D], Engine.State[D])] =
+    mapEvalState[Event[D], Engine.State[D], (Event[D], Engine.State[D])](input, qs, (e: Event[D]) => runE(e).run)
+
+  // Functions for type bureaucracy
+
+  private def pure[A](a: A): HandleP[A, D] = Applicative[({type L[T] = HandleP[T, D]})#L].pure(a)
+
+  private val unit: HandleP[Unit, D] = pure(())
+
+  private val get: HandleP[Engine.State[D], D] =
+    MonadState[({type L[T] = Handle[T, D]})#L, Engine.State[D]].get.toHandleP
+
+  private def gets[A](f: (Engine.State[D]) => A): HandleP[A, D] =
+    MonadState[({type L[T] = Handle[T, D]})#L, Engine.State[D]].gets(f).toHandleP
+
+  private def modify(f: (Engine.State[D]) => Engine.State[D]): HandleP[Unit, D] =
+    MonadState[({type L[T] = Handle[T, D]})#L, Engine.State[D]].modify(f).toHandleP
+
+  private def getS(id: Sequence.Id): HandleP[Option[Sequence.State], D] = get.map(_.sequences.get(id))
+
+  private def getSs[A](id: Sequence.Id)(f: Sequence.State => A): HandleP[Option[A], D] =
+    gets(_.sequences.get(id).map(f))
+
+  private def modifyS(id: Sequence.Id)(f: Sequence.State => Sequence.State): HandleP[Unit, D] =
+    modify(
+      st => Engine.State(
+        st.userData,
+        st.conditions,
+        st.operator,
+        st.sequences.get(id).map(
+          s => st.sequences.updated(id, f(s))).getOrElse(st.sequences)
+      )
+    )
+
+  private def putS(id: Sequence.Id)(s: Sequence.State): HandleP[Unit, D] =
+    modify(st => Engine.State[D](st.userData, st.conditions, st.operator, st.sequences.updated(id, s)))
+
+  // For debugging
+  def printSequenceState(id: Sequence.Id): HandleP[Unit, D] =
+    getSs(id)((qs: Sequence.State) => Task.now(println(qs)).liftM[({type L[M[_], A] = HandleStateT[M, A, D]})#L]).void // scalastyle:ignore
+
+}
+
+object Engine {
 
-  final case class State(conditions: Conditions, operator: Option[Operator], sequences: Map[Sequence.Id, Sequence.State])
+  final case class State[D](userData: D, conditions: Conditions, operator: Option[Operator], sequences: Map[Sequence.Id, Sequence.State])
```

Comment:
Will `userData` contain the calibrations queue?

Reply:
Do we really need to be polymorphic here? I suppose in most cases the `UserData` will be a well known structure. I question this due to the complexity introduced.

Also, should/could we put conditions/operator inside `D`?

Reply:
Yes, `userData` contains the calibration queue, and anything else we could need in the future. I want to put conditions/operator inside, but for that we need a mechanism to pass them to inside the Actions.